libraries/libfile/ZipArchive.cpp

```diff
@@ -0,0 +1,343 @@
+#include <libfile/ZipArchive.h>
+#include <libsystem/Logger.h>
+#include <libsystem/compression/Deflate.h>
+#include <libsystem/compression/Inflate.h>
+#include <libutils/BinaryReader.h>
+#include <libutils/Endian.h>
+
+// Central header
+#define ZIP_END_OF_CENTRAL_DIR_HEADER_SIG 0x06054b50
+#define ZIP_CENTRAL_DIR_HEADER_SIG 0x02014b50
+
+// Local header
+#define ZIP_LOCAL_DIR_HEADER_SIG 0x04034b50
+
+enum ExtraFieldType : uint16_t
+{
+    EFT_ZIP64 = 0x0001,
+    EFT_EXTENDED_TIMESTAMP = 0x5455,
+    EFT_NEW_UNIX = 0x7875,
+};
+
+using le_eft = LittleEndian<ExtraFieldType>;
+
+enum EntryFlags : uint16_t
+{
+    EF_NONE = 0,
+    EF_ENCRYPTED = 1 << 0,
+    EF_COMPRESSION_OPTION = 1 << 1,
+    EF_COMPRESSION_OPTION2 = 1 << 2,
+    EF_DATA_DESCRIPTOR = 1 << 3,
+};
+
+using le_flags = LittleEndian<EntryFlags>;
+
+enum CompressionMethod : uint16_t
+{
+    CM_UNCOMPRESSED = 0,
+    CM_SHRUNK = 1,
+    CM_DEFLATED = 8,
+};
+
+using le_compression = LittleEndian<CompressionMethod>;
+
+struct __packed CentralDirectoryFileHeader
+{
+    le_uint32_t signature;
+    le_uint16_t version;
+    le_uint16_t version_required;
+    le_flags flags;
+    le_compression compression;
+    le_uint16_t mod_time;
+    le_uint16_t mod_date;
+    le_uint32_t crc;
+    le_uint32_t compressed_size;
+    le_uint32_t uncompressed_size;
+    le_uint16_t len_filename;
+    le_uint16_t len_extrafield;
+    le_uint16_t len_comment;
+    le_uint16_t disk_start;
+    le_uint16_t internal_attr;
+    le_uint32_t external_attr;
+    le_uint32_t local_header_offset;
+};
+
+struct __packed CentralDirectoryEndRecord
+{
+    le_uint32_t signature;
+    le_uint16_t disk1;
+    le_uint16_t disk2;
+    le_uint16_t disk_entries;
+    le_uint16_t total_entries;
+    le_uint32_t central_dir_size;
+    le_uint32_t central_dir_offset;
+    le_uint16_t len_comment;
+};
+
+struct __packed LocalHeader
+{
+    le_uint32_t signature;
+    le_uint16_t version;
+    le_flags flags;
+    le_compression compression;
+    le_uint16_t mod_time;
+    le_uint16_t mod_date;
+    le_uint32_t crc;
+    le_uint32_t compressed_size;
+    le_uint32_t uncompressed_size;
+    le_uint16_t len_filename;
+    le_uint16_t len_extrafield;
+};
+
+static_assert(sizeof(LocalHeader) == 30, "LocalHeader has invalid size!");
+
+struct __packed DataDescriptor
+{
+    le_uint32_t crc;
+    le_uint32_t compressed_size;
+    le_uint32_t uncompressed_size;
+};
+
+static_assert(sizeof(DataDescriptor) == 12, "DataDescriptor has invalid size!");
+
+ZipArchive::ZipArchive(File file, bool read) : Archive(file)
+{
+    if (read)
+    {
+        read_archive();
+    }
+}
+
+void ZipArchive::read_archive()
+{
+    _valid = false;
+    Result result = _file.read_all((void **)&_data, &_size);
```

Comment:
The **Slice** would not transfer ownership correctly when i was storing it inside a variable. See the chat we had in Discord, but it would be better indeed to get that version working.